### DISCUSSION TITLE
Data exchange with OPC server on PC

### DISCUSSION THREAD

**Discussion by AlexArGC** [eModbus/eModbus#263] [Q&A] [answered]
Am I doing the right thing?
I need to send data from ESP32 sensors to an OPC server (server mode) and then to SCADA...
ESP32 and OPC are on the same network, firewalls are disabled.

**ESP32 -> ModBUS TCP -> OPC -> SCADA**

![image](https://user-images.githubusercontent.com/48761297/215456760-0c2694be-e010-464e-9940-873b936d81f4.png)

![image](https://user-images.githubusercontent.com/48761297/215457688-46455c03-fcfb-49c1-b7f1-cf8b950a670f.png)

I took an example from TCP03async, but I don't understand how to put registers on the bus?
I am missing the WiFi network settings in the above code, because everything is alright with them.

```
#include <ModbusClientTCPasync.h>     

IPAddress ipMB = {192, 168, 0, 50};   
uint16_t portMB = 1502;               
ModbusClientTCPasync MB(ipMB, portMB); 

void handleData(ModbusMessage response, uint32_t token);
void handleError(Error error, uint32_t token);
void ModBusRequest();

MB.onDataHandler(&handleData);
MB.onErrorHandler(&handleError);
MB.setTimeout(10000);
MB.setIdleTimeout(60000);

setup () 
{
  timer.setInterval(3 * 1000, ModBusRequest);
}

loop ()
{
  timer.run()
}

void handleData(ModbusMessage response, uint32_t token)
{
  Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", response.getServerID(), response.getFunctionCode(), token, response.size());
  for (auto &byte : response)
  {
    Serial.printf("%02X ", byte);
  }
  Serial.println("");
}

void handleError(Error error, uint32_t token)
{
  ModbusError me(error);
  Serial.printf("Error response: %02X - %s token: %d\n", (int)me, (const char *)me, token);
}

void ModBusRequest()
{
  static unsigned long eTokenMB = 0;
  eTokenMB = millis();

  Serial.printf("sending request with token %d\n", (uint32_t)eTokenMB);
  Error err;
  err = MB.addRequest((uint32_t)eTokenMB, 1, WRITE_HOLD_REGISTER, 0, 1024);
  if (err != SUCCESS)
  {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

}
```

To serial I get send:
```
sending request with token 114568
sending request with token 117568
sending request with token 120566
sending request with token 123564
Error response: E0 - Timeout token: 114568
sending request with token 126570
Error response: E0 - Timeout token: 117568
sending request with token 129568
Error response: E0 - Timeout token: 120566
sending request with token 132564
```

I don't understand something or I'm doing something wrong.

**Comment by Miq1** [comment]
If your OPC is defined as server ID 1 and will accept a write request on address 0, the ESP code looks okay. Your timeout is set to 10s, but you are sending every 3s. That may be intended and is not wrong, but you will send two more requests before a timeout for the first can be noticed.

Since the error code is E0 (TIMEOUT) and no connection error like EA, a TCP connection seems to have been established,only the server is not responding over it. I cannot read your cyrillic screen shots, so you may have to investigate yourself, why it is not.

**Reply by AlexArGC** [reply]
Thanks for the answer! I will investigate further. I thought I was doing something wrong with your examples.

1. Yes, correct device ID on OPC server ID = 1.
![изображение](https://user-images.githubusercontent.com/48761297/215486455-81bd9ece-5ce7-453b-b221-340952a5c850.png)
2. The OPC communication channel is configured as a server, on port 1502
![изображение](https://user-images.githubusercontent.com/48761297/215486890-b139ac48-c42f-4441-937c-0e33fce7e8fc.png)
3. Am I doing the right thing by adding registers to the ModBUS stack on the OPC server, and then trying to change the registers from the ModBUS TCP client to the ESP32?
4. The OPC has a timeout of 100ms, changing to 3000ms did nothing.
5. I'll try to enable detailed logs on ESP32...
6. In the OPC server diagnostics, I receive the following messages:
```
30.01.2023 18:17:00.008 - *:1502;REC;12;00 03 00 00 00 06 01 06 00 00 04 00
30.01.2023 18:16:57.125 - *:1502;SEND;12;00 00 00 00 00 06 01 03 00 00 00 01
30.01.2023 18:16:57.005 - *:1502;REC;12;00 02 00 00 00 06 01 06 00 00 04 00
30.01.2023 18:16:54.110 - *:1502;SEND;12;00 00 00 00 00 06 01 03 00 00 00 01
30.01.2023 18:16:54.006 - *:1502;REC;12;00 01 00 00 00 06 01 06 00 00 04 00
30.01.2023 18:16:51.092 - *:1502;SEND;12;00 00 00 00 00 06 01 03 00 00 00 01
30.01.2023 18:16:51.082 - *:1502;REC;12;00 00 00 00 00 06 01 06 00 00 04 00
30.01.2023 18:16:51.070 - *:1502;SEND;12;00 00 00 00 00 06 01 03 00 00 00 01
```

**Reply by Miq1** [reply]
Looks *almost* OK 😀 
You are receiving 
```
00 01 00 00 00 06 01 06 00 00 04 00
```
from the ESP, that is 
- transaction ID 0x0001
- protocol ID 0x0000
- payload length 6
- (TCP header complete and OK)
- server ID 1
- function code WRITE_HOLD_REGISTER 0x06
- address 0x0000
- new content 0x0400, that is 1024

So the request is all fine. Now the response:
```
00 00 00 00 00 06 01 03 00 00 00 01
```
- transaction ID 0x0000 (!!! - see below)
- protocol ID 0x0000
- payload length 6
- server ID 1
- function code 0x03 READ_HOLD_REGISTER (!!!)
- address 0x0000
- length 1

Your OPC is behaving as a client itself! Therefore the transaction ID does not match that of the response, plus, it tries to read data from the ESP, which is a client itself and will not respond at all.

You will have to decide which way to go:
- either have the ESP as a client (like it is now) and write data into registers of the OPC acting as a proper server - it will have to be configured differently of course
- or make the ESP a server and provide a server function to respond to the OPC's READ_HOLD_REGISTER requests with data.

**Reply by AlexArGC** [reply]
> Your OPC is behaving as a client itself! Therefore the transaction ID does not match that of the response, plus, it tries to read data from the ESP, which is a client itself and will not respond at all.
> 
> You will have to decide which way to go:
> 
>     * either have the ESP as a client (like it is now) and write data into registers of the OPC acting as a proper server - it will have to be configured differently of course
> 
>     * or make the ESP a server and provide a server function to respond to the OPC's READ_HOLD_REGISTER requests with data.

Thank you. You are right as always! You have a great library. As a wish, if it is possible to supplement the help with various examples :-)

I tested your example with another simulator utility. And there the register is correctly written to the bus. I will look for a working OPS server with slave mode.

There is confusion with the register address, all programs process them differently, deviating from the standard :-(.

![изображение](https://user-images.githubusercontent.com/48761297/215849176-aff6cca0-c120-4478-a750-c6d95714d1ad.png)
![изображение](https://user-images.githubusercontent.com/48761297/215849202-db676759-95be-4287-965e-c43715f017ab.png)

```
// =================================================================================================
// eModbus: Copyright 2020 by Michael Harwerth, Bert Melis and the contributors to ModbusClient
//               MIT license - see license.md for details
// =================================================================================================

// Example code to show the usage of the eModbus library. 
// Please refer to root/Readme.md for a full description.

// Includes: <Arduino.h> for Serial etc., WiFi.h for WiFi support
#include <Arduino.h>
#include <WiFi.h>

// Include the header for the ModbusClient TCP style
#include "ModbusClientTCPasync.h"

#ifndef MY_SSID
#define MY_SSID "WIFIRS96"
#endif
#ifndef MY_PASS
#define MY_PASS "11223344"
#endif

char ssid[] = MY_SSID;                     // SSID and ...
char pass[] = MY_PASS;                     // password for the WiFi network used
IPAddress ip = {192, 168, 0, 50};          // IP address of modbus server
uint16_t port = 502;                      // port of modbus server

// Create a ModbusTCP client instance
ModbusClientTCPasync MB(ip, port);

// Define an onData handler function to receive the regular responses
// Arguments are Modbus server ID, the function code requested, the message data and length of it, 
// plus a user-supplied token to identify the causing request
void handleData(ModbusMessage response, uint32_t token) 
{
  Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", response.getServerID(), response.getFunctionCode(), token, response.size());
  for (auto& byte : response) {
    Serial.printf("%02X ", byte);
  }
  Serial.println("");
}

// Define an onError handler function to receive error responses
// Arguments are the error code returned and a user-supplied token to identify the causing request
void handleError(Error error, uint32_t token) 
{
  // ModbusError wraps the error code and provides a readable error message for it
  ModbusError me(error);
  Serial.printf("Error response: %02X - %s token: %d\n", (int)me, (const char *)me, token);
}

// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Connect to WiFi
  WiFi.begin(ssid, pass);
  delay(200);
  while (WiFi.status() != WL_CONNECTED) {
    Serial.print(". ");
    delay(1000);
  }
  IPAddress wIP = WiFi.localIP();
  Serial.printf("WIFi IP address: %u.%u.%u.%u\n", wIP[0], wIP[1], wIP[2], wIP[3]);

// Set up ModbusTCP client.
// - provide onData handler function
  MB.onDataHandler(&handleData);
// - provide onError handler function
  MB.onErrorHandler(&handleError);
// Set message timeout to 2000ms and interval between requests to the same host to 200ms
  MB.setTimeout(10000);
// Start ModbusTCP background task
  MB.setIdleTimeout(60000);
  MB.connect();

}

// loop() - nothing done here today!
void loop() {
  static unsigned long lastMillis = 0;
  if (millis() - lastMillis > 5000) {
    lastMillis = millis();

    // Create request for
    // (Fill in your data here!)
    // - server ID = 1
    // - function code = 0x06 (цфкшеу holding register)
    // - start address to read = word 10
    // - number of words to read = 4
    // - token to match the response with the request. We take the current millis() value for it.
    //
    // If something is missing or wrong with the call parameters, we will immediately get an error code 
    // and the request will not be issued
    Serial.printf("sending request 0х06 with token %d\n", (uint32_t)lastMillis);
    Error err;
    err = MB.addRequest((uint32_t)lastMillis, 1, WRITE_HOLD_REGISTER, 10, (uint16_t)millis());
    if (err != SUCCESS) {
      ModbusError e(err);
      Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
      Serial.printf("Erroer count = %d, message count = %d\n", MB.getErrorCount(), MB.getMessageCount());
    }

 
    // Else the request is processed in the background task and the onData/onError handler functions will get the result.
    //
    // The output on the Serial Monitor will be (depending on your WiFi and Modbus the data will be different):
    //     __ OK __
    //     . WIFi IP address: 192.168.178.74
    //     Response: serverID=20, FC=3, Token=0000056C, length=11:
    //     14 03 04 01 F6 FF FF FF 00 C0 A8
  }
}
```

**Reply by Miq1** [reply]
There is a common habit to map registers to some higher addresses, like 40000 and above for hold registers etc. This is not part of the standard and not required for any technical reasons. You may of course simulate that by checking the addresses to be in the range, but your internal location or processing is totally up to you.

Good luck!

**Comment by AlexArGC** [comment]
I found an OPC server that worked in slave mode!

Suddenly someone will come in handy for experiments https://opc-server.com/catalog/modbusopc/modbus_opc/

![изображение](https://user-images.githubusercontent.com/48761297/215878000-7dea07a7-82fe-458c-82b0-ef96d90112a7.png)

**Reply by AlexArGC** [reply]
**How to send float type to OPC server.**

I understand that in `addRequest()` it requires the entire package to be completely formed.

One of the ways that seemed to me the most simple. Might be useful for newbies. We use the `union` structure.

```
union uUnion
{
  float dFloat;
  uint16_t word[2];
};
...
 static unsigned long eTokenMB = 0;
 eTokenMB = millis();

 uUnion uTemp;
 uTemp.dFloat = (float)Temp1;
 err = MB.addRequest((uint32_t)eTokenMB, 1, WRITE_MULT_REGISTERS, 0, 2, 4, uTemp.word);
...

```

**Reply by Miq1** [reply]
That way you will send the float in the byte and word order of the machine you are currently on. The ``add()`` function in the library takes care to send a float in MSB IEEE754 order to keep it portable.

**Comment by AlexArGC** [comment]
**Error: queue is full**

Why is there an error? What is the limit? `addToQueue()` is called every 3 seconds.

```
sending request with token 33039000
[1;33m[E] 3303900| ModbusClientTCPasync.cpp [ 160] addToQueue: queue is full
[0mError creating request: E8 - Request queue full
```

I noticed that when the WiFi connection is broken, the objects do not close. Of course, I added a connection check, but this does not give a 100% guarantee.

```
sending request with token 6722849
sending request with token 6722856
sending request with token 6722859
[673564][E][WiFiMulti.cpp:117] run(): [WIFI] no networks found
[673564][E][WiFiMulti.cpp:191] run(): [WIFI] no matching wifi found!
sending request with token 6752848
sending request with token 6752854
sending request with token 6752858
sending request with token 6782849
sending request with token 6782856
sending request with token 6782859
```

**Reply by Miq1** [reply]
The possible frequency of requests mostly is limited by the ability of a server to recover from a previous response. 3 seconds is a lot, but you should consider it will take much longer for a broken connection to time out (f.i. 30s is a typical internal time of the ESP8266). In the meantime you will have fired 9 more requests that will have been placed in the queue. Depending on the queue limit the queue will quickly fill up.

**Comment by AlexArGC** [comment, marked as answer]
Yes, I understand that. What settings in your library can be changed to close connections faster? How to increase the request quota? I have ESP32...

**Reply by Miq1** [reply]
There is none AFAIK unfortunately. The connection timeout is set internally "under the hood" of the ESP in the Espressif firmware. We are struggling with it ourselves currently on the ESP8266. 
Why is your WiFi dropping out? Perhaps you could tackle that first?

**Comment by AlexArGC** [comment]
Thank you for your answers!

The fact is that the OPC server is located remotely (on a white IP) and communication breaks are possible. Therefore, I deliberately emulated the state of an unstable WiFi. I wanted to check the stability of the library.

Added a connection check before requests. So far it's worked for days without a hitch.

```
  wifiConnected = (WiFi.status() == WL_CONNECTED);
  if (wifiConnected)
  {
```

**P.S.**
I'm doing a little experiment...
